_posts/2016-07-09-making-dnssec-work-for-an-aws-hosted-saas.markdown

```diff
@@ -0,0 +1,190 @@
+
+[search.usa.gov](http://search.digitalgov.gov/) is a SaaS that serves thousand of other government agencies. During our recent project to migrate the search.usa.gov server infrastructure to Amazon Web Services' (AWS) Elastic Compute Cloud (EC2), we were presented with a challenging combination of requirements whose solution seems worth sharing:
+
+1. The search.usa.gov EC2 infrastructure needed to support the addition, removal, or replacement of servers on-the-fly without interruption of service
+1. The search.usa.gov customers - other government agencies - needed to be able to direct their users to the SaaS offering either at `search.usa.gov`, or at "masked" domain name such as `search.someagency.gov`
+1. The DNS for all `search.usa.gov` records were [required to be signed by DNSSEC](https://www.whitehouse.gov/sites/default/files/omb/memoranda/fy2008/m08-23.pdf)
+
+As we'll see, this seemed at first to be a "pick two out of three" situation. Satisfying all three requirements at once required a spiffy open-source software feature and a two-tier DNS server architecture that we'll describe in this post.
+
+### Wildcard Records and Another Level of Indirection
+
+> All problems in computer science can be solved by another level of indirection, except of course for the problem of too many indirections. - [David Wheeler](https://en.wikipedia.org/wiki/David_Wheeler_(British_computer_scientist))
+
+The simplest of our three requirements was to allow customers to choose whether to use the `search.usa.gov` website directly, or whether to create a "masked" domain name such as `search.someagency.gov`. This was entirely in the customer's control; they could simply create a [CNAME](https://en.wikipedia.org/wiki/CNAME_record) in their own DNS zone to point `search.someagency.gov` to `search.usa.gov` if they wanted a masked domain name.
+
+We knew that we wanted to avoid a "Big Bang" migration event in which all of our customers were suddenly pointed to the new AWS hosting infrastructure, so we went one step further by asking search.usa.gov customers to CNAME their agency-specific hostnames to _customer-specific_ CNAMEs. This would allow us to shift traffic around on a customer-by-customer basis - both to the new infrastructure, but also back to the old infrastructure if the need arose:
+
+```
+$ORIGIN someagency.gov.
+search IN CNAME someagency.sites.infr.search.usa.gov
+```
+
+Once a fair number of our customers had these CNAMEs in place, we were able to switch them over to our new AWS infrastructure on a customer-by-customer basis rather than switching everyone at once:
```

Comment:
nit: the previous sentence also uses "on a customer-by-customer basis" - maybe i'm being too strict here though